src/log-store/src/raft_engine/backend.rs

```diff
@@ -0,0 +1,548 @@
+
+        let mut prev = None;
+        let engine = self.engine.read().unwrap();
+        if prev_kv {
+            prev = engine_get(&engine, &key)?;
+        }
+        engine_put(&engine, key, value)?;
+        Ok(PutResponse { prev_kv: prev })
+    }
+
+    async fn batch_put(&self, req: BatchPutRequest) -> Result<BatchPutResponse, Self::Error> {
+        let BatchPutRequest { kvs, prev_kv } = req;
+        let mut batch = LogBatch::with_capacity(kvs.len());
+
+        let mut prev_kvs = if prev_kv {
+            Vec::with_capacity(kvs.len())
+        } else {
+            vec![]
+        };
+
+        for kv in kvs {
+            if prev_kv && let Some(kv) = engine_get(&self.engine.read().unwrap(), &kv.key)? {
+                prev_kvs.push(kv);
+            }
+            batch
+                .put(SYSTEM_NAMESPACE, kv.key, kv.value)
+                .context(RaftEngineSnafu)?;
+        }
+
+        self.engine
+            .read()
+            .unwrap()
+            .write(&mut batch, false)
+            .context(RaftEngineSnafu)?;
+
+        Ok(BatchPutResponse { prev_kvs })
+    }
+
+    async fn batch_get(&self, req: BatchGetRequest) -> Result<BatchGetResponse, Self::Error> {
+        let mut response = BatchGetResponse {
+            kvs: Vec::with_capacity(req.keys.len()),
+        };
+        let engine = self.engine.read().unwrap();
+        for key in req.keys {
+            let value = engine.get(SYSTEM_NAMESPACE, &key).unwrap();
+            response.kvs.push(KeyValue { key, value });
+        }
+        Ok(response)
+    }
+
+    async fn compare_and_put(
+        &self,
+        req: CompareAndPutRequest,
+    ) -> Result<CompareAndPutResponse, Self::Error> {
+        let CompareAndPutRequest { key, expect, value } = req;
+
+        let mut batch = LogBatch::with_capacity(1);
+        let engine = self.engine.write().unwrap();
+        let existing = engine_get(&engine, &key)?;
+        let eq = existing
+            .as_ref()
+            .map(|kv| kv.value == expect)
+            .unwrap_or(false);
+        if eq {
+            batch
+                .put(SYSTEM_NAMESPACE, key, value)
+                .context(RaftEngineSnafu)?;
+            engine.write(&mut batch, false).context(RaftEngineSnafu)?;
+        }
+        Ok(CompareAndPutResponse {
+            success: eq,
+            prev_kv: existing,
+        })
+    }
+
+    async fn delete_range(
+        &self,
+        req: DeleteRangeRequest,
+    ) -> Result<DeleteRangeResponse, Self::Error> {
+        let DeleteRangeRequest {
+            key,
+            range_end,
+            prev_kv,
+        } = req;
+
+        let range = RangeRequest {
+            key,
+            range_end,
+            limit: 0,
+            keys_only: true,
+        };
+        let range_resp = self.range(range).await?;
+
+        let mut prev_kvs = vec![];
+        let mut deleted = 0;
+        for kv in range_resp.kvs {
+            self.delete(&kv.key, false).await?;
```

Comment:
I think it's better not call `self.delete` here. Acquired too many times of locks.

Reply:
fixed in [6276174](https://github.com/GreptimeTeam/greptimedb/pull/2280/commits/6276174017099321622fa692fe9cd1fb32b47ab0#diff-74b285f40ce368ca6fe6df02183ca1e98c68112554e907aa9e7271ee6ca2d6feR254-R257)